test/Microsoft.ML.Tests/Scenarios/Api/CrossValidation.cs

```diff
@@ -0,0 +1,81 @@
+        /// a data source (optionally with stratification column), come up with an instantiable transform
+        /// and trainer pipeline, and it will handle (1) splitting up the data, (2) training the separate
+        /// pipelines on in-fold data, (3) scoring on the out-fold data, (4) returning the set of
+        /// evaluations and optionally trained pipes. (People always want metrics out of xfold,
+        /// they sometimes want the actual models too.)
+        /// </summary>
+        [Fact]
+        void CrossValidation()
+        {
+            var dataPath = GetDataPath(SentimentDataPath);
+            var testDataPath = GetDataPath(SentimentTestPath);
+
+            int numFolds = 5;
+            using (var env = new TlcEnvironment(seed: 1, conc: 1))
+            {
+                // Pipeline.
+                var loader = new TextLoader(env, MakeSentimentTextLoaderArgs(), new MultiFileSource(dataPath));
+
+                var trans = TextTransform.Create(env, MakeSentimentTextTransformArgs(false), loader);
+                var random = new GenerateNumberTransform(env, trans, "StratificationColumn");
+                // Train.
+                var trainer = new LinearClassificationTrainer(env, new LinearClassificationTrainer.Arguments
+                {
+                    NumThreads = 1,
+                    ConvergenceTolerance = 1f
+                });
+
+                // Auto-caching.
+                IDataView trainData = trainer.Info.WantCaching ? (IDataView)new CacheDataView(env, random, prefetch: null) : random;
+                var metrics = new List<BinaryClassificationMetrics>();
+                for (int fold = 0; fold < numFolds; fold++)
+                {
+                    var trainFilter = new RangeFilter(env, new RangeFilter.Arguments()
+                    {
+                        Column = "StratificationColumn",
+                        Min = (Double)fold / numFolds,
+                        Max = (Double)(fold + 1) / numFolds,
+                        Complement = true
+                    }, trainData);
+
+                    // Auto-normalization.
+                    var trainRoles = new RoleMappedData(trainFilter, label: "Label", feature: "Features");
+                    NormalizeTransform.CreateIfNeeded(env, ref trainRoles, trainer);
+
+                    var predictor = trainer.Train(new Runtime.TrainContext(trainRoles));
+                    var testFilter = new RangeFilter(env, new RangeFilter.Arguments()
+                    {
+                        Column = "StratificationColumn",
+                        Min = (Double)fold / numFolds,
+                        Max = (Double)(fold + 1) / numFolds,
+                        Complement = false
+                    }, trainData);
+                    // Auto-normalization.
+                    var testRoles = new RoleMappedData(testFilter, label: "Label", feature: "Features");
```

Comment:
>testRoles [](start = 24, length = 9)

I think this is wrong here. We should not re-normalize test data, we should apply existing normalizer (line 56) to test data

Reply:
Still incorrect. Now we normalize on the entire dataset, which means that we also use test data for training a normalizer. This is incorrect.
You should 'normalize if needed' on the train set, as before, and then 'apply all transforms' to the test set.

---
In reply to: [208314510](https://github.com/dotnet/machinelearning/pull/653#discussion_r208314510) [](ancestors = 208314510)

Reply:
After this one is resolved, I'm fine

---
In reply to: [208616551](https://github.com/dotnet/machinelearning/pull/653#discussion_r208616551) [](ancestors = 208616551,208314510)